src/io_notifier.cpp

```diff
@@ -0,0 +1,256 @@
+#include "coro/io_notifier.hpp"
+
+#include <array>
+#include <stdexcept>
+
+namespace coro
+{
+
+io_notifier::io_notifier()
+#if defined(__FreeBSD__) || defined(__APPLE__) || defined(__OpenBSD__) || defined(__NetBSD__)
+    : m_fd{::kqueue()}
+#elif defined(__linux__)
+    : m_fd{::epoll_create1(EPOLL_CLOEXEC)}
+#endif
+{
+}
+
+io_notifier::~io_notifier()
+{
+    if (m_fd != -1)
+    {
+        ::close(m_fd);
+        m_fd = -1;
+    }
+}
+
+timer_handle::timer_handle(const void* timer_handle_ptr, io_notifier& notifier)
+#if defined(__FreeBSD__) || defined(__APPLE__) || defined(__OpenBSD__) || defined(__NetBSD__)
+    : m_fd{1},
+#elif defined(__linux__)
+    : m_fd(::timerfd_create(CLOCK_MONOTONIC, TFD_NONBLOCK | TFD_CLOEXEC)),
+#endif
+      m_timer_handle_ptr(timer_handle_ptr)
+{
+#if defined(__FreeBSD__) || defined(__APPLE__) || defined(__OpenBSD__) || defined(__NetBSD__)
+    (void)notifier;
+#elif defined(__linux__)
+    notifier.watch(m_fd, coro::poll_op::read, const_cast<void*>(m_timer_handle_ptr), true);
+#endif
+}
+
+auto io_notifier::watch_timer(const timer_handle& timer, std::chrono::nanoseconds duration) -> bool
+{
+#if defined(__FreeBSD__) || defined(__APPLE__) || defined(__OpenBSD__) || defined(__NetBSD__)
```

Comment:
its possible `duration` has a negative value here, can you handle that just like the linux branch does?